File: ex153.py
class Solution:
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        for i in range(0, len(nums) - 1):
            if nums[i + 1] < nums[i]:   return nums[i + 1]
            
        return nums[0]

sol = Solution()
print(sol.findMin([4,5,6,7,0,1,2]))

File: code/ex153.cpp
// Binary Search

class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size();
    
        int lo = 0, hi = n - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if ((mid + 1 >= n || nums[mid + 1] > nums[mid]) && (mid - 1 < 0 || nums[mid - 1] > nums[mid]))
                return nums[mid];
            else if (nums[mid] > nums[hi])
                lo = mid + 1;
            else
                hi = mid - 1;
        }
        
        return -1;
    }
};